arch/cortex-m/src/mpu.rs

```diff
@@ -150,35 +150,31 @@ impl MPU {
 /// The cortex-m MPU has eight regions, all of which must be configured (though
 /// unused regions may be configured as disabled). This struct caches the result
 /// of region configuration calculation
-pub struct CortexMConfig {
+pub struct CortexMConfig<const NUM_REGIONS: usize> {
     /// The computed region configuration for this process.
-    regions: [CortexMRegion; 8],
+    regions: [CortexMRegion; NUM_REGIONS],
     /// Has the configuration changed since the last time the this process
     /// configuration was written to hardware?
     is_dirty: Cell<bool>,
 }
 
 const APP_MEMORY_REGION_NUM: usize = 0;
 
-impl Default for CortexMConfig {
-    fn default() -> CortexMConfig {
-        CortexMConfig {
-            regions: [
-                CortexMRegion::empty(0),
-                CortexMRegion::empty(1),
-                CortexMRegion::empty(2),
-                CortexMRegion::empty(3),
-                CortexMRegion::empty(4),
-                CortexMRegion::empty(5),
-                CortexMRegion::empty(6),
-                CortexMRegion::empty(7),
-            ],
+impl<const NUM_REGIONS: usize> Default for CortexMConfig<NUM_REGIONS> {
+    fn default() -> Self {
+        // a bit of a hack to initialize array without unsafe
```

Comment:
Is there any way to impl for a specific value? i.e. create impl's for `CortexMConfig<8>` and `CortexMConfig<16>`, which covers our use cases, and the would throw a 'missing impl' error for any other value of `NUM_REGIONS`? Could address this.

Reply:
Unfortunately I don't think we can get this to work using just `min_const_generics`. Implementing for just 8 and 16 leads to the following error on Imix (which uses `NUM_REGIONS = 8`):
```
error[E0277]: the trait bound `CortexMConfig<NUM_REGIONS>: Default` is not satisfied
   --> arch/cortex-m/src/mpu.rs:392:5
    |
392 |     type MpuConfig = CortexMConfig<NUM_REGIONS>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `CortexMConfig<NUM_REGIONS>`
    |
   ::: kernel/src/platform/mpu.rs:94:21
    |
94  |     type MpuConfig: Default + Display;
    |                     ------- required by this bound in `kernel::mpu::MPU::MpuConfig`
    |
    = help: the following implementations were found:
              <CortexMConfig<16_usize> as Default>
              <CortexMConfig<8_usize> as Default>
```

Adding a blanket implementation on top of those two that panics gives the following error:

```
error[E0119]: conflicting implementations of trait `core::default::Default` for type `mpu::CortexMConfig<8_usize>`:
   --> arch/cortex-m/src/mpu.rs:169:1
    |
163 | impl<const NUM_REGIONS: usize> Default for CortexMConfig<NUM_REGIONS> {
    | --------------------------------------------------------------------- first implementation here
...
169 | impl Default for CortexMConfig<8> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `mpu::CortexMConfig<8_usize>`
```

and I do not think that min_const_generics offers a way to `impl<const NUM_REGIONS: usize> ... where NUM_REGIONS != 8`

Reply:
Ah, yeah, that makes sense. The compiler can't know at that point which versions it will eventually need. Perhaps someday as const generics evolves that can be optimized, but I don't think a few extra cycles on startup is going to hurt much for today.